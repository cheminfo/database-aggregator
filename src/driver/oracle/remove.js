'use strict';

const sourceSequence = require('../../mongo/models/sourceSequence');
const model = require('../../mongo/model');
const debug = require('../../util/debug')('driver:oracle');

const common = require('./common');

module.exports = function (options) {
  common.checkOptions(['connectString', 'query'], options);
  return doDelete(options);
};

async function doDelete(options) {
  const collection = options.collection;
  const Model = model.getSource(collection);

  // connect to Oracle
  const oracleConn = await common.connect(options);

  let query = options.query;

  query = `SELECT id FROM (\n${query}\n) inner_table`;

  const sourceIds = await common.getIDs(oracleConn, query);
  await oracleConn.release();

  // Element that are found in copied source but not in original source ought to be deleted
  // We do this by setting data to null, so that aggregation knows about the deletion
  const copiedIds = await Model.find({ data: { $ne: null } }, { _id: 1 })
    .lean()
    .exec();
  const idsToDelete = new Set();

  for (let i = 0; i < copiedIds.length; i++) {
    let id = copiedIds[i]._id;
    if (!sourceIds.has(id)) {
      idsToDelete.add(id);
    }
  }
  const removeThreshold =
    options.removeThreshold === undefined ? 0.01 : options.removeThreshold;
  const percentToDelete = idsToDelete.size / copiedIds.length;
  if (percentToDelete > removeThreshold) {
    debug.warn(
      `removal of data from ${collection} cancelled (maximum: ${removeThreshold *
        100}%, actual: ${percentToDelete * 100}%)`
    );
    return;
  }

  for (const id of idsToDelete) {
    debug.trace(`delete ${id} from ${collection}`);
    await Model.findByIdAndUpdate(id, {
      $set: {
        data: null,
        date: new Date(),
        sequentialID: await sourceSequence.getNextSequenceID(
          `source_${collection}`
        )
      }
    }).exec();
  }
}
