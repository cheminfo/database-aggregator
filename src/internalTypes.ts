import { Document } from 'mongoose';

export interface ISchedulerLogDocument extends Document {
  taskId: string;
  pid: string;
  state: ISchedulerStatus[];
  date: Date;
}

export interface ISchedulerStatus {
  status: string;
  date: Date;
  stdout?: string;
  stderr?: string;
  message?: string;
}

export interface ISourceDocument<T extends object = object> extends Document {
  id: string;
  commonID: string;
  sequentialID: number;
  date: Date;
  data: T;
}

export interface IAggregationDocument<T = any> extends Document {
  id: string;
  date: Date;
  value: T;
}

export interface ISourceSequenceDocument extends Document {
  _id: string;
  seq: number;
  version: number;
}

export interface IAggregationSequenceDocument extends Document {
  _id: string;
  sourceSeq: {
    [key: string]: number;
  };
}

export interface IScheduleDefinition {
  id: string;
  worker: string;
  immediate: boolean;
  cronRule?: string;
  deps: string[];
  noConcurrency: string[];
  arg: string;
  type: string;
}
