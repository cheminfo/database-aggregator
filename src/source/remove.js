'use strict';

const getDriverFunction = require('./getDriverFunction');

function remove(options) {
  const driverRemove = getDriverFunction(options.driver, 'remove');
  return driverRemove;
}

module.exports = remove;
