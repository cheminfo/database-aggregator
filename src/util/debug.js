'use strict';

const Debug = require('debug');

var error = Debug('aggregate:error');
var warn = Debug('aggregate:warn');
var debug = Debug('aggregate:debug');
var trace = Debug('aggregate:trace');

warn.log = console.warn.bind(console);
debug.log = console.log.bind(console);
trace.log = console.log.bind(console);

module.exports = function (prefix) {
  const func = (message) => debug(`(${prefix}) ${message}`);
  func.error = (message) => error(`(${prefix}) ${message}`);
  func.warn = (message) => warn(`(${prefix}) ${message}`);
  func.debug = func;
  func.trace = (message) => trace(`(${prefix}) ${message}`);
  return func;
};
