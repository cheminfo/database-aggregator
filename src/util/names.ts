const AGGREGATION = 'aggregation';
const SOURCE = 'source';
const META = 'meta';

const REMOVE_ID = 'source_remove_';
const COPY_ID = 'source_copy_';
const COPY_MISSING_ID = 'source_copy_missing_ids_';

export function getModelName(prefix: string, name: string) {
  return `${prefix}_${name}`;
}

export function getSourceName(name: string) {
  return getModelName(SOURCE, name);
}

export function getMetaModelName(name: string) {
  return getModelName(META, name);
}

export function getSourceModelName(name: string) {
  return getModelName(SOURCE, name);
}

export function getAggregationModelName(name: string) {
  return getModelName(AGGREGATION, name);
}

export function getCopyTaskId(collection: string) {
  return COPY_ID + collection;
}

export function getRemoveTaskId(collection: string) {
  return REMOVE_ID + collection;
}

export function getCopyMissingIdTaskId(collection: string) {
  return COPY_MISSING_ID + collection;
}
