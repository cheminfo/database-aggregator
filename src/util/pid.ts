/* eslint-disable no-process-exit */

const path = require("path");
const fs = require("fs");

const isRunning = require("is-running");

const pidFile = path.join(__dirname, "../../source.pid");

export function start() {
  try {
    const pid = fs.readFileSync(pidFile, "utf8");
    if (isRunning(pid)) {
      console.log(`process is running (pid: ${pid})`);
      process.exit(0);
    }
    console.log("pid file is here but process died");
    fs.unlinkSync(pidFile);
  } catch (e) {
    if (e.code === "ENOENT") {
      console.log("pid file does not exist yet");
    } else {
      throw e;
    }
  }
  const currentPid = process.pid;
  fs.writeFileSync(pidFile, String(currentPid));
}

export function stop(code?: number) {
  code = code || 0;
  fs.unlinkSync(pidFile);
  process.exit(code);
}
