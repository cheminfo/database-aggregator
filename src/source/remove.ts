import { connect } from '../mongo/connection';
import { getSource } from '../mongo/model';

import { getNextSequenceID } from '../mongo/models/sourceSequence';
import { ISourceConfigElement } from '../types';
import { debugUtil } from '../util/debug';
import { getDriver } from './getDriver';

const debug = debugUtil('source:remove');

export async function remove(config: ISourceConfigElement) {
  const driver = getDriver(config.driver);

  // Get complete list of source Ids
  let sourceIds = await driver.getIds(config.driverConfig);
  if (!(sourceIds instanceof Set)) {
    sourceIds = new Set(sourceIds);
  }

  const collection = config.collection;
  const Model = getSource(collection);

  await connect();

  // Element that are found in copied source but not in original source ought to be deleted
  // We do this by setting data to null, so that aggregation knows about the deletion
  const copiedIds = await Model.find({ data: { $ne: null } }, { id: 1 })
    .lean()
    .exec();
  const idsToDelete = new Set();

  for (const { id } of copiedIds) {
    if (!sourceIds.has(id)) {
      idsToDelete.add(id);
    }
  }
  const removeThreshold =
    config.removeThreshold === undefined ? 0.01 : config.removeThreshold;
  const percentToDelete = idsToDelete.size / copiedIds.length;
  if (percentToDelete > removeThreshold) {
    debug.warn(
      `removal of data from ${collection} cancelled (maximum: ${removeThreshold *
        100}%, actual: ${percentToDelete * 100}%)`
    );
    return;
  }

  for (const id of idsToDelete) {
    debug.trace(`delete ${id} from ${collection}`);
    await Model.updateOne(
      { id },
      {
        $set: {
          data: null,
          date: new Date(),
          sequentialID: await getNextSequenceID(collection)
        }
      }
    ).exec();
  }
}
