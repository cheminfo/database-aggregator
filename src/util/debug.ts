import * as Debug from "debug";

let error = Debug("aggregator:error");
let warn = Debug("aggregator:warn");
let debug = Debug("aggregator:debug");
let trace = Debug("aggregator:trace");

warn.log = console.warn.bind(console);
debug.log = console.log.bind(console);
trace.log = console.log.bind(console);

export function debugUtil(prefix: string) {
  return {
    error: (message: string) => error(`(${prefix}) ${message}`),
    warn: (message: string) => warn(`(${prefix}) ${message}`),
    debug: (message: string) => debug(`(${prefix}) ${message}`),
    trace: (message: string) => trace(`(${prefix}) ${message}`),
  };
}
