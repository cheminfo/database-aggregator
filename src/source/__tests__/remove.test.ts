import {
  connect,
  disconnect,
  dropDatabase,
  insertData
} from '../../../test/mongoSetup';
import { getCollection, clean } from '../../../test/util';
import { remove } from '../remove';
import { ISourceConfigElement } from '../../internalTypes';

beforeAll(connect);
afterAll(disconnect);

const collection = getCollection('source_test');
const config: ISourceConfigElement = {
  driver: {
    getIds: (driverConfig) => {
      const data = ['test1', 'test3'];
      if (driverConfig.type === 'set') {
        return new Set(data);
      }
      return data;
    },
    getData: (x, y, z) => {
      // void
    }
  },
  driverConfig: {},
  collection: 'test'
};

const testData = {
  meta_source_sequence: [
    {
      _id: 'test',
      seq: 3
    }
  ],
  source_test: [
    {
      id: 'test1',
      commonID: 'test1',
      date: new Date(0),
      sequentialID: 1,
      data: {}
    },
    {
      id: 'test2',
      commonID: 'test2',
      date: new Date(0),
      sequentialID: 2,
      data: {}
    },
    {
      id: 'test3',
      commonID: 'test3',
      date: new Date(0),
      sequentialID: 3,
      data: {}
    }
  ]
};

describe('source remove', () => {
  beforeEach(async () => {
    await dropDatabase();
    await insertData(testData);
  });
  it('should ignore remove if threshold is too low', async () => {
    await remove(config);
    const data = await collection.find().toArray();
    expect(data[1].data).toEqual({});
    expect(clean(data)).toMatchSnapshot();
  });

  it('should remove if threshold is high enough', async () => {
    await remove(Object.assign(config, { removeThreshold: 0.9 }));
    const data = await collection.find().toArray();
    expect(data[1].data).toBeNull();
    delete data[1].date;
    expect(clean(data)).toMatchSnapshot();
  });

  it('should remove when driver returns a set', async () => {
    await remove(
      Object.assign(config, {
        driverConfig: { type: 'set' },
        removeThreshold: 0.9
      })
    );
    const data = await collection.find().toArray();
    expect(data[1].data).toBeNull();
    delete data[1].date;
    expect(clean(data)).toMatchSnapshot();
  });
});
