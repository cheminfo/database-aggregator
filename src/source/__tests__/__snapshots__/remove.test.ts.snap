// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`source remove should ignore remove if threshold is too low 1`] = `
Array [
  Object {
    "commonID": "test1",
    "data": Object {},
    "date": 1970-01-01T00:00:00.000Z,
    "id": "test1",
    "sequentialID": 1,
  },
  Object {
    "commonID": "test2",
    "data": Object {},
    "date": 1970-01-01T00:00:00.000Z,
    "id": "test2",
    "sequentialID": 2,
  },
  Object {
    "commonID": "test3",
    "data": Object {},
    "date": 1970-01-01T00:00:00.000Z,
    "id": "test3",
    "sequentialID": 3,
  },
]
`;

exports[`source remove should remove if threshold is high enough 1`] = `
Array [
  Object {
    "commonID": "test1",
    "data": Object {},
    "date": 1970-01-01T00:00:00.000Z,
    "id": "test1",
    "sequentialID": 1,
  },
  Object {
    "commonID": "test2",
    "data": null,
    "id": "test2",
    "sequentialID": 4,
  },
  Object {
    "commonID": "test3",
    "data": Object {},
    "date": 1970-01-01T00:00:00.000Z,
    "id": "test3",
    "sequentialID": 3,
  },
]
`;

exports[`source remove should remove when driver returns a set 1`] = `
Array [
  Object {
    "commonID": "test1",
    "data": Object {},
    "date": 1970-01-01T00:00:00.000Z,
    "id": "test1",
    "sequentialID": 1,
  },
  Object {
    "commonID": "test2",
    "data": null,
    "id": "test2",
    "sequentialID": 4,
  },
  Object {
    "commonID": "test3",
    "data": Object {},
    "date": 1970-01-01T00:00:00.000Z,
    "id": "test3",
    "sequentialID": 3,
  },
]
`;
